pkg/controllers/cinder/cinder.go

```diff
@@ -0,0 +1,151 @@
+	// Minimal interval between controller resyncs. The controller will detect
+	// new share types in Cinder and create StorageClasses for them at least
+	// once per this interval.
+	resyncInterval = 20 * time.Minute
+
+	operatorConditionPrefix = "CinderController"
+)
+
+func NewCinderController(
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	informers v1helpers.KubeInformersForNamespaces,
+	openStackClient *openStackClient,
+	csiControllers []Runnable,
+	eventRecorder events.Recorder) factory.Controller {
+
+	scInformer := informers.InformersFor("").Storage().V1().StorageClasses()
+	c := &CinderController{
+		operatorClient:     operatorClient,
+		kubeClient:         kubeClient,
+		storageClassLister: scInformer.Lister(),
+		openStackClient:    openStackClient,
+		csiControllers:     csiControllers,
+		eventRecorder:      eventRecorder.WithComponentSuffix("CinderController"),
+	}
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(operatorClient).ResyncEvery(resyncInterval).WithInformers(
+		operatorClient.Informer(),
+		scInformer.Informer(),
+	).ToController("CinderController", eventRecorder)
+}
+
+func (c *CinderController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(4).Infof("Cinder sync started")
+	defer klog.V(4).Infof("Cinder sync finished")
+
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorv1.Managed {
+		return nil
+	}
+
+	volumeTypes, err := c.openStackClient.GetVolumeTypes()
+	if err != nil {
+		return err
+	}
+
+	if len(volumeTypes) == 0 {
+		err = fmt.Errorf("no volume types found for Cinder service")
+		klog.V(2).Error(err, "No volume types found for Cinder service")
+		return err
+	}
+
+	if !c.controllersRunning {
+		klog.V(4).Infof("Starting CSI driver controllers")
+		for _, ctrl := range c.csiControllers {
+			go ctrl.Run(ctx, 1)
+		}
+		c.controllersRunning = true
+	}
+	err = c.syncStorageClasses(ctx, volumeTypes)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *CinderController) syncStorageClasses(ctx context.Context, volumeTypes []volumetypes.VolumeType) error {
+	var errs []error
+	for _, volumeType := range volumeTypes {
+		klog.V(4).Infof("Syncing storage class for volumeType type %s", volumeType.Name)
+		sc := c.generateStorageClass(volumeType)
+		_, _, err := resourceapply.ApplyStorageClass(c.kubeClient.StorageV1(), c.eventRecorder, sc)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errors.NewAggregate(errs)
+}
+
+func (c *CinderController) applyStorageClass(ctx context.Context, expected *storagev1.StorageClass) error {
+	_, _, err := resourceapply.ApplyStorageClass(c.kubeClient.StorageV1(), c.eventRecorder, expected)
+	return err
+}
+
+func (c *CinderController) generateStorageClass(volumeType volumetypes.VolumeType) *storagev1.StorageClass {
+	storageClassName := util.StorageClassNamePrefix + strings.ToLower(volumeType.Name)
```

Comment:
Seems like the answer to my question is no. There is [code to validate a name](https://github.com/kubernetes/apimachinery/blob/74cf9d365d5e3eee61ff26e62d90a0cffd9f2ba8/pkg/util/validation/validation.go#L198) (thanks, @JoelSpeed), but nothing to automatically fix it up. I just had a quick look in cinder and afaict there is no validation on name beyond it being a string. It's length doesn't even appear to be well defined (which almost certainly means it's db-dependent), so we should assume it's the wild west.

I don't think it would be correct to raise an error here if the OpenStack object name is not a valid metadata.name, so we're going to have to translate. I'll have a think.